delira/training/predictor.py

```diff
@@ -0,0 +1,208 @@
+                 prepare_batch_fn=lambda *x: x, **kwargs):
+        """
+
+        Parameters
+        ----------
+        model :
+            model to predict from; must be callable
+
+        """
+        self._setup(model, convert_batch_to_npy_fn, prepare_batch_fn, **kwargs)
+
+    def _setup(self, network, convert_batch_to_npy_fn, prepare_batch_fn):
+        self.module = network
+        self._convert_batch_to_npy_fn = convert_batch_to_npy_fn
+        self._prepare_batch = prepare_batch_fn
+
+    def __call__(self, data):
+        return self.predict(data)
+
+    def predict(self, data, keys=["data"]):
+        data = self._prepare_batch(data)
+
+        if isinstance(data, dict):
+            data = [data[key] for key in keys]
+
+        return self._convert_batch_to_npy_fn(
+            self.module(
+                *data
+            )
+        )
+
+    def predict_data_mgr(self, datamgr, batchsize=None, verbose=False):
+        """
+        Defines a routine to predict data obtained from a batchgenerator
+
+        Parameters
+        ----------
+        batchgen : MultiThreadedAugmenter
+            Generator Holding the Batches
+        batchsize : Artificial batchsize (sampling will be done with batchsize
+                    1 and sampled data will be stacked to match the artificial
+                    batchsize)(default: None)
+
+        Raises
+        ------
+        NotImplementedError
+            If not overwritten by subclass
+
+        """
+
+        orig_num_aug_processes = datamgr.n_process_augmentation
+        orig_batch_size = datamgr.batch_size
+
+        datamgr.batch_size = 1
+        datamgr.n_process_augmentation = 1
+
+        batchgen = datamgr.get_batchgen()
+
+        predictions_all, inputs_all = [], []
+
+        n_batches = batchgen.generator.num_batches * batchgen.num_processes
+
+        if verbose:
+            iterable = tqdm(enumerate(batchgen), unit=' sample',
+                            total=n_batches, desc='Test')
+
+        else:
+            iterable = enumerate(batchgen)
+
+        batch_list = []
+
+        for i, batch in iterable:
+
+            if not batch_list and (n_batches - i) < batchsize:
+                batchsize = n_batches - i
+                logger.debug("Set Batchsize down to %d to avoid cutting "
+                             "of the last batches" % batchsize)
+
+            batch_list.append(batch)
+
+            # if queue is full process queue:
+            if batchsize is None or len(batch_list) >= batchsize:
+
+                batch_dict = {}
+                for batch in batch_list:
+                    for key, val in batch.items():
+                        if key in batch_dict.keys():
+                            batch_dict[key].append(val)
+                        else:
+                            batch_dict[key] = [val]
+
+                for key, val_list in batch_dict.items():
+                    batch_dict[key] = np.array(val_list)
```

Comment:
why this got changed from [np.concatenate](https://github.com/justusschock/delira/blob/b663d982d87f3a0be7b346d25576770b184d4db9/delira/training/tf_trainer.py#L463) to np.array? should fix the singleton issue presumably:

Reply:
My mistake. Will fix this probably tomorrow